### PR TITLE
Stencil support

### PR DESCRIPTION
**Pull request by marcolanaro**
Issue: #4600 

## What I did
- Added support for Stencil.
- I added example kitchen sink for Stencil.

## How to test

- Is this testable with Jest or Chromatic screenshots? No
- Does this need a new example in the kitchen sink apps? Yes
- Does this need an update to the documentation? No

### CODE REVIEW

**Comment by shilman** [comment]
@marcolanaro thanks so much for putting this PR together--this is a great start!

I have some questions to help me wrap my head around it.

In the issue you mention that this is a half-solution and you still need to run `npm run build.watch` by hand. do you envision the full solution being one where storybook effectively runs that on your behalf, and then cleans up gracefully on shutdown?

If you're up for it, I'd love to chat more about this on Discord https://discord.gg/storybook

**Comment by marcolanaro** [comment]
> @marcolanaro thanks so much for putting this PR together--this is a great start!
> 
> I have some questions to help me wrap my head around it.
> 
> In the issue you mention that this is a half-solution and you still need to run `npm run build.watch` by hand. do you envision the full solution being one where storybook effectively runs that on your behalf, and then cleans up gracefully on shutdown?
> 
> If you're up for it, I'd love to chat more about this on Discord https://discord.gg/storybook

@shilman This is correct. This would unblock stencil developers writing stories easily.

I saw many different works around that:
- vanilla javascript (quite bloated if you have to pass down rich data or event listeners)
- output jsdoc to feed to lit and then use of `lit-html` (another template language to learn) for rendering in the stories
- additional proxy server that replace storybook iframe
- [use](https://dev.to/ofhouse/enhance-your-stencil-web-components-in-storybook-with-knobs-actions-and-jsx-54m4) of `jsx-dom` (not yet production ready).

This would enable developers to stay in the same environment and write stories as they write components. Yes, it's not perfect from storybook point of view.
To be able to run the compiler using storybook, I need to do a lot of investigations. Stencil expose a very limited set of APIs and it's time consuming looking at the internals. 
This PR took me an entire weekend and I will not have much more spare time in the next month.
I can look into the compiler in August (so I would say let's catch up then for phase 2 if it's not a problem for you).

**Comment by shilman** [comment]
@marcolanaro thanks for getting back to me. just let me know when you have more time to discuss in August. In the meantime we'll take a look at your solution and see whether we can build on it

File: examples/stencil-kitchen-sink/package.json [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    "test.watch": "stencil test --spec --e2e --watchAll",
+    "generate": "stencil generate",
+    "storybook": "start-storybook -p 9001"
+  },
+  "dependencies": {
+    "@stencil/core": "^2.0.1"
+  },
+  "devDependencies": {
+    "@storybook/source-loader": "*",
+    "@storybook/stencil": "../../app/stencil",
```
Comment (merceyz) [inline examples/stencil-kitchen-sink/package.json:30]: This will actually pack that folder into the cache, you should use a version instead

**Comment by sawden** [comment]
Is the stencil support still in progress? 🙂

**Comment by cymptom** [comment]
I'd also love to know if/when official Stencil support is likely to arrive. For our Stencil-based design system, we've been holding off on Storybook until official support arrives.

**Comment by julzmon** [comment]
I was hoping to see some progress when coming here but sad to see no :(
Are we waiting for something official from Stencil team, or folks have to time to work on it?

**Comment by shilman** [comment]
I haven't heard back from @marcolanaro about continuing this work.

In the meantime I have discussed with the Stencil core team and they are looking into how they might provide official support from their end 🤞 Please upvote the issues here https://github.com/ionic-team/stencil/issues/3104

**Comment by splitinfinities** [comment]
Thank you so much @shilman! 

Yes, the Stencil team is looking for feedback on the linked issue above. I am working through the logistics of what it means for Stencil as a compiler that can not only produce web components, but can also produce components for other frameworks. So any other related feedback and suggestions on direction is welcome!

**Comment by ndelangen** [comment]
I'm closing this due to inactivity, sorry.

**Pull request closed without merging**